File: tv-fork-patches/patched/common/src/main/java/com/liskovsoft/smartyoutubetv2/common/litetube/LitetubeApi.java
package com.liskovsoft.smartyoutubetv2.common.litetube;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import org.json.JSONArray;
import org.json.JSONObject;
import java.util.concurrent.TimeUnit;

/**
 * Lightweight Litetube REST client. Uses its own OkHttp singleton.
 */
public final class LitetubeApi {
    private static final MediaType JSON = MediaType.parse("application/json");

    private static final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(15, TimeUnit.SECONDS)
            .build();

    private LitetubeApi() {}

    public static JSONObject startDevice(String apiBase) {
        return postJson(apiBase + "/api/devices/start", new JSONObject(), null);
    }

    public static JSONObject pollDevice(String apiBase, String code) {
        return getJson(apiBase + "/api/devices/poll?code=" + code, null, 35);
    }

    public static JSONArray fetchProxyPool(String apiBase, String jwt) {
        return getJsonArray(apiBase + "/api/proxy/pool", jwt, 15);
    }

    private static JSONObject getJson(String url, String bearer, long timeoutSec) {
        try {
            Request req = baseRequest(url, bearer).get().build();
            OkHttpClient timeoutClient = client.newBuilder()
                    .callTimeout(timeoutSec, TimeUnit.SECONDS)
                    .readTimeout(timeoutSec, TimeUnit.SECONDS)
                    .build();
            Response resp = timeoutClient.newCall(req).execute();
            try {
                if (!resp.isSuccessful()) return null;
                String body = resp.body() != null ? resp.body().string() : "";
                return new JSONObject(body);
            } finally {
                resp.close();
            }
        } catch (Exception e) {
            return null;
        }
    }

    private static JSONArray getJsonArray(String url, String bearer, long timeoutSec) {
        try {
            Request req = baseRequest(url, bearer).get().build();
            OkHttpClient timeoutClient = client.newBuilder()
                    .callTimeout(timeoutSec, TimeUnit.SECONDS)
                    .readTimeout(timeoutSec, TimeUnit.SECONDS)
                    .build();
            Response resp = timeoutClient.newCall(req).execute();
            try {
                if (!resp.isSuccessful()) return null;
                String body = resp.body() != null ? resp.body().string() : "";
                return new JSONArray(body);
            } finally {
                resp.close();
            }
        } catch (Exception e) {
            return null;
        }
    }

    private static JSONObject postJson(String url, JSONObject body, String bearer) {
        try {
            Request req = baseRequest(url, bearer)
                    .post(RequestBody.create(JSON, body.toString()))
                    .build();
            Response resp = client.newCall(req).execute();
            try {
                if (!resp.isSuccessful()) return null;
                String txt = resp.body() != null ? resp.body().string() : "";
                return new JSONObject(txt);
            } finally {
                resp.close();
            }
        } catch (Exception e) {
            return null;
        }
    }

    private static Request.Builder baseRequest(String url, String bearer) {
        Request.Builder b = new Request.Builder().url(url)
                .header("Accept", "application/json")
                .header("User-Agent", "LitetubeTV/0.1 AndroidTV");
        if (bearer != null && !bearer.isEmpty())
            b.header("Authorization", bearer);
        return b;
    }
}

File: tv-fork-patches/patched/common/src/main/java/com/liskovsoft/smartyoutubetv2/common/litetube/LitetubeApi.kt
package com.liskovsoft.smartyoutubetv2.common.litetube

import com.liskovsoft.sharedutils.okhttp.OkHttpManager
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Lightweight Litetube REST client. Reuses the shared OkHttp singleton so it
 * picks up whatever the system-proxy selector just wrote to http(s).proxyHost.
 */
object LitetubeApi {
    private const val TAG = "LitetubeApi"
    private val JSON = "application/json".toMediaTypeOrNull()

    fun startDevice(apiBase: String): JSONObject? = postJson(
        apiBase + "/api/devices/start",
        body = JSONObject(),
        bearer = null,
    )

    /** Long-poll up to server-side DEVICE_POLL_MAX_SEC. Returns null on timeout. */
    fun pollDevice(apiBase: String, code: String): JSONObject? = getJson(
        apiBase + "/api/devices/poll?code=$code", bearer = null,
        timeoutSec = 35, // server caps at 30s; add slack for TLS handshake
    )

    fun fetchProxyPool(apiBase: String, jwt: String): JSONArray? = getJsonArray(
        apiBase + "/api/proxy/pool", bearer = jwt, timeoutSec = 15,
    )

    // ---- low-level helpers ---------------------------------------------

    private fun getJson(url: String, bearer: String?, timeoutSec: Long): JSONObject? {
        val req = baseRequest(url, bearer).get().build()
        return runCatching {
            OkHttpManager.instance().okHttpClient.newBuilder()
                .callTimeout(timeoutSec, TimeUnit.SECONDS)
                .readTimeout(timeoutSec, TimeUnit.SECONDS)
                .build()
                .newCall(req).execute().use { resp ->
                    val body = resp.body?.string().orEmpty()
                    if (!resp.isSuccessful) null else JSONObject(body)
                }
        }.getOrNull()
    }

    private fun getJsonArray(url: String, bearer: String?, timeoutSec: Long): JSONArray? {
        val req = baseRequest(url, bearer).get().build()
        return runCatching {
            OkHttpManager.instance().okHttpClient.newBuilder()
                .callTimeout(timeoutSec, TimeUnit.SECONDS)
                .readTimeout(timeoutSec, TimeUnit.SECONDS)
                .build()
                .newCall(req).execute().use { resp ->
                    val body = resp.body?.string().orEmpty()
                    if (!resp.isSuccessful) null else JSONArray(body)
                }
        }.getOrNull()
    }

    private fun postJson(url: String, body: JSONObject, bearer: String?): JSONObject? {
        val req = baseRequest(url, bearer)
            .post(body.toString().toRequestBody(JSON))
            .build()
        return runCatching {
            OkHttpManager.instance().okHttpClient.newCall(req).execute().use { resp ->
                val txt = resp.body?.string().orEmpty()
                if (!resp.isSuccessful) null else JSONObject(txt)
            }
        }.getOrNull()
    }

    private fun baseRequest(url: String, bearer: String?): Request.Builder {
        val b = Request.Builder().url(url)
            .header("Accept", "application/json")
            .header("User-Agent", "LitetubeTV/0.1 AndroidTV")
        if (!bearer.isNullOrBlank()) b.header("Authorization", bearer)
        return b
    }
}
